File: src/priv/bajdcc/LALR1/interpret/module/std/ModuleStdBase.java
package priv.bajdcc.LALR1.interpret.module.std;

import priv.bajdcc.LALR1.grammar.Grammar;
import priv.bajdcc.LALR1.grammar.runtime.IRuntimeDebugInfo;
import priv.bajdcc.LALR1.grammar.runtime.RuntimeCodePage;
import priv.bajdcc.LALR1.interpret.module.IInterpreterModule;
import priv.bajdcc.util.ResourceLoader;

/**
 * 【模块】基础库
 *
 * @author bajdcc
 */
public class ModuleStdBase implements IInterpreterModule {

	private static ModuleStdBase instance = new ModuleStdBase();
	private RuntimeCodePage runtimeCodePage;

	public static ModuleStdBase getInstance() {
		return instance;
	}

	@Override
	public String getModuleName() {
		return "std.base";
	}

	@Override
	public String getModuleCode() {
		return ResourceLoader.load(getClass());
	}

	@Override
	public RuntimeCodePage getCodePage() throws Exception {
		if (runtimeCodePage != null)
			return runtimeCodePage;

		String base = ResourceLoader.load(getClass());

		Grammar grammar = new Grammar(base);
		RuntimeCodePage page = grammar.getCodePage();
		IRuntimeDebugInfo info = page.getInfo();

		return runtimeCodePage = page;
	}
}

File: src/main/java/com/bajdcc/LALR1/interpret/module/std/ModuleStdBase.kt
package com.bajdcc.LALR1.interpret.module.std

import com.bajdcc.LALR1.grammar.Grammar
import com.bajdcc.LALR1.grammar.runtime.RuntimeCodePage
import com.bajdcc.LALR1.interpret.module.IInterpreterModule
import com.bajdcc.util.ResourceLoader

/**
 * 【模块】基础库
 *
 * @author bajdcc
 */
class ModuleStdBase : IInterpreterModule {
    private var runtimeCodePage: RuntimeCodePage? = null

    override val moduleName: String
        get() = "std.base"

    override val moduleCode: String
        get() = ResourceLoader.load(javaClass)

    override val codePage: RuntimeCodePage
        @Throws(Exception::class)
        get() {
            if (runtimeCodePage != null)
                return runtimeCodePage!!

            val base = ResourceLoader.load(javaClass)

            val grammar = Grammar(base)
            val page = grammar.codePage

            runtimeCodePage = page
            return page
        }

    companion object {

        val instance = ModuleStdBase()
    }
}